src/allmydata/web/check_results.py

```diff
@@ -531,119 +671,167 @@ def json(self, ctx):
+            return tag
         return ""
 
-    def data_pre_repair_problems(self, ctx, data):
+    @renderer
+    def pre_repair_problems(self, req, tag):
         all_objects = self.monitor.get_status().get_all_results()
+        problems = []
+
         for path in sorted(all_objects.keys()):
             r = all_objects[path]
             assert ICheckAndRepairResults.providedBy(r)
             cr = r.get_pre_repair_results()
             if not cr.is_healthy():
-                yield path, cr
-
-    def render_problem(self, ctx, data):
-        path, cr = data
-        return ctx.tag[self._join_pathstring(path), ": ",
-                       self._html(cr.get_summary())]
-
-    def render_post_repair_problems_p(self, ctx, data):
-        c = self.monitor.get_status().get_counters()
-        if (c["count-objects-unhealthy-post-repair"]
-            or c["count-corrupt-shares-post-repair"]):
-            return ctx.tag
+                problem = self._join_pathstring(path), ": ", self._html(cr.get_summary())
+                problems.append({"problem": problem})
+
+        return SlotsSequenceElement(tag, problems)
+
+    @renderer
+    def post_repair_problems_p(self, req, tag):
+        if (self._get_counter("count-objects-unhealthy-post-repair")
+            or self._get_counter("count-corrupt-shares-post-repair")):
+            return tag
         return ""
 
-    def data_post_repair_problems(self, ctx, data):
+    @renderer
+    def post_repair_problems(self, req, tag):
         all_objects = self.monitor.get_status().get_all_results()
+        problems = []
+
         for path in sorted(all_objects.keys()):
             r = all_objects[path]
             assert ICheckAndRepairResults.providedBy(r)
             cr = r.get_post_repair_results()
             if not cr.is_healthy():
-                yield path, cr
+                problem = self._join_pathstring(path), ": ", self._html(cr.get_summary())
+                problems.append({"problem": problem})
 
-    def render_servers_with_corrupt_shares_p(self, ctx, data):
-        if self.monitor.get_status().get_counters()["count-corrupt-shares-pre-repair"]:
-            return ctx.tag
-        return ""
-    def data_servers_with_corrupt_shares(self, ctx, data):
-        return [] # TODO
-    def render_server_problem(self, ctx, data):
-        pass
+        return SlotsSequenceElement(tag, problems)
 
-
-    def render_remaining_corrupt_shares_p(self, ctx, data):
-        if self.monitor.get_status().get_counters()["count-corrupt-shares-post-repair"]:
-            return ctx.tag
+    @renderer
+    def servers_with_corrupt_shares_p(self, req, tag):
+        if self._get_counter("count-corrupt-shares-pre-repair"):
+            return tag
         return ""
-    def data_post_repair_corrupt_shares(self, ctx, data):
-        return [] # TODO
 
-    def render_share_problem(self, ctx, data):
-        pass
+    @renderer
+    def servers_with_corrupt_shares(self, req, tag):
```

Comment:
Isn't this a duplicate method to the one above?

Reply:
They look very similar, and they probably use the same `Monitor` to get their data.  I will need to manually test this (I don't think this is covered by tests), or extend test coverage.

It is possible that `DeepCheckResultsRendererElement` and `DeepCheckAndRepairResultsRendererElement` can share `servers_with_corrupt_shares()`, but I am unsure how to do that.  Copy-pasting the same thing across two classes isn't very appealing.  Maybe `servers_with_corrupt_shares()` can be pulled into a common base class?

Reply:
Probably not the best solution, but implementing `DeepCheckAndRepairResultsRendererElement` as a `DeepCheckResultsRendererElement` (uhh, sorry about those really silly class names) allows for some code sharing between those two classes, including `servers_with_corrupt_shares`.

However `post_repair_corrupt_shares()` still remains un-implemented.  I have filed ticket [3371](https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3371) for that.